lib/restrictor.cu

```diff
@@ -1,145 +1,109 @@
 #include <color_spinor_field.h>
-#include <tune_quda.h>
-#include <launch_kernel.cuh>
-
-#include <jitify_helper.cuh>
+#include <tunable_block_reduction.h>
 #include <kernels/restrictor.cuh>
 
 namespace quda {
 
-  template <typename Float, typename vFloat, int fineSpin, int fineColor, int coarseSpin, int coarseColor,
-            int coarse_colors_per_thread>
-  class RestrictLaunch : public Tunable {
+  struct Aggregates {
+    // List of block sizes we wish to instantiate.  The required block
+    // size is equal to number of fine points per aggregate, rounded
+    // up to a whole power of two.  So for example, 2x2x2x2 and
+    // 3x3x3x1 aggregation would both use the same block size 32
+#ifndef QUDA_FAST_COMPILE_REDUCE
+    static constexpr std::array<unsigned int, 6> block = {32, 64, 128, 256, 512, 1024};
```

Comment:
Is there a reason for powers of 2 as opposed to all multiples of `32`, or something more like multiples of `64` after `64`, etc? I'm just wondering if there's a performance loss here due to outright unused warps in some cases.

Reply:
There's no performance loss.  I did a fairly thorough sweep, and the restrictor implementation in GK is always faster than in develop (which has MG aggregate size = CUDA thread block size).

Typically, the autotuner prefers to run with either one or two CTAs per block, to limit the occupancy, so any unused threads don't really cause any overhead.